File: src/luhn.ts
/**
 * Generic Luhn checksum validation.
 *
 * Validates any digit string using the Luhn
 * algorithm (ISO/IEC 7812-1). No length
 * restriction. For credit card validation
 * (13-19 digits), use `creditcard` instead.
 */

import { luhnValidate } from "#checksums/luhn";
import { clean } from "#util/clean";
import { err } from "#util/result";
import { isdigits } from "#util/strings";

import type { ValidateResult, Validator } from "./types";

const compact = (value: string): string =>
  clean(value, " -.");

const validate = (value: string): ValidateResult => {
  const v = compact(value);
  if (v.length < 1) {
    return err("INVALID_LENGTH", "Value must not be empty");
  }
  if (!isdigits(v)) {
    return err(
      "INVALID_FORMAT",
      "Value must contain only digits",
    );
  }
  if (!luhnValidate(v)) {
    return err(
      "INVALID_CHECKSUM",
      "Luhn check digit does not match",
    );
  }
  return { valid: true, compact: v };
};

const format = (value: string): string => compact(value);

/** Generic Luhn Validator. */
const luhn: Validator = {
  name: "Luhn",
  localName: "Luhn",
  abbreviation: "Luhn",
  entityType: "any",
  examples: ["4111111111111111", "18"] as const,
  compact,
  format,
  validate,
};

export default luhn;
export { compact, format, validate };

File: crates/stdnum-core/src/validators/global/luhn.rs
use crate::{
  compact_without, decimal_digits_strict, is_ascii_digits, luhn_checksum,
  random_digits,
  types::{
    EntityType, ValidationError, ValidationResult, Validator, ValidatorScope,
    ValidatorSpec,
  },
};

pub static VALIDATOR: Validator = Validator::new(ValidatorSpec {
  id: "luhn",
  name: "Luhn",
  local_name: "Luhn",
  abbreviation: "Luhn",
  aliases: &["Luhn", "Luhn algorithm", "mod 10"],
  candidate_pattern: r"\d{2,}",
  scope: ValidatorScope::Global,
  entity_type: EntityType::Any,
  source_url: Some("https://en.wikipedia.org/wiki/Luhn_algorithm"),
  lengths: &[],
  examples: &["4111111111111111", "18"],
  compact,
  format,
  validate,
  generate: Some(generate_default),
  parse: None,
});

#[must_use]
pub fn compact(value: &str) -> String {
  compact_without(value, &[' ', '-', '.'])
}

#[must_use]
pub fn format(value: &str) -> String {
  compact(value)
}

pub fn validate(value: &str) -> ValidationResult {
  let value = compact(value);
  if value.is_empty() {
    return Err(ValidationError::InvalidLength("Value must not be empty"));
  }
  if !is_ascii_digits(&value) {
    return Err(ValidationError::InvalidFormat(
      "Value must contain only digits",
    ));
  }
  if luhn_checksum(&decimal_digits_strict(&value)) != 0 {
    return Err(ValidationError::InvalidChecksum(
      "Luhn check digit does not match",
    ));
  }
  Ok(value)
}

#[must_use]
pub fn generate(length: usize) -> String {
  let payload_length = length.saturating_sub(1);
  let payload = random_digits(payload_length);
  let mut partial = decimal_digits_strict(&payload);
  partial.push(0);
  let check = 10_u32
    .saturating_sub(luhn_checksum(&partial))
    .rem_euclid(10);
  format!("{payload}{check}")
}

fn generate_default() -> String {
  generate(16)
}
